### PR TITLE
WooCommerce: Add variation editing UI.

### LINKED ISSUE #12770: WooCommerce: Add/Edit Product: Variations Display & Form Inputs
Display a product's variations as rows in a table under the "variation types" component. Should include inputs for managing price, dimensions/weight, and stock. An edit icon should also be present to open a editing modal (see #13512 for modal implementation).

Previously, See #12760 and #12684.

### PR DESCRIPTION
**Pull request by justinshreve**
This adds the variations list table and UI for managing variation data like price, dimensions, weight, and stock. Closes #12770.

To test:
* Visit `http://calypso.localhost:3000/store/products/:site/add`
* Start typing in some variation types.
* Type in and toggle variation fields and make sure their values update.

screen shot 2017-05-09 at 8 48 52 am

### CODE REVIEW

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+					<FormTextInput
+						name="width"
+						placeholder={ translate( 'W' ) }
+						type="number"
+						value={ dimensions.width || '' }
+						onChange={ this.setDimension }
+					/>
+					<FormTextInputWithAffixes
+						name="height"
+						placeholder={ translate( 'H' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **8** times:
`translate( 'h' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+						currencySymbolPrefix="$"
+						name="price"
+						value={ variation.regular_price || '' }
+						onChange={ this.setPrice }
+					/>
+				</td>
+				<td>
+					<FormTextInput
+						name="length"
+						placeholder={ translate( 'L' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'L', { context: 'momentjs format string (sameElse)'} )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variations-table.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	render() {
+		const { variations, translate } = this.props;
+		return (
+			<table className="products__product-form-variation-table">
+				<thead>
+					<tr>
+						<th></th>
+						<th></th>
+						<th>{ translate( 'Price' ) }</th>
+						<th>{ translate( 'Dimensions & weight' ) }</th>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: Hi! I've found a possible matching string that has already been translated **77** times:
`translate( 'Dimensions' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-variation-rows)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by kellychoffman** [comment]
Did a first pass at styling. Will need to do more tomorrow. 

2 requests: 1. Can we pull out the product styles into their own stylesheet (like we did for shipping settings in https://github.com/Automattic/wp-calypso/pull/13604/commits/7c91e04b803289a27d7a8f7a0bb2128c3f4797b7) 2. Can we add the full width layout to this page? @marcinbot was able to add it to the Shipping Settings page, but it has since been removed and rebased so I'm not sure how it was done.

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+						name="width"
+						placeholder={ translate( 'W' ) }
+						type="number"
+						value={ dimensions.width || '' }
+						onChange={ this.setDimension }
+						className="products__product-dimensions-width"
+					/>
+					<FormTextInputWithAffixes
+						name="height"
+						placeholder={ translate( 'H' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **8** times:
`translate( 'h' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+						name="price"
+						value={ variation.regular_price || '' }
+						onChange={ this.setPrice }
+						size="4"
+					/>
+				</td>
+				<td className="products__product-dimensions-input">
+					<FormTextInput
+						name="length"
+						placeholder={ translate( 'L' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'L', { context: 'momentjs format string (sameElse)'} )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variations-table.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		const { variations, translate } = this.props;
+		return (
+			<div className="products__product-form-variation-table-wrapper">
+				<table className="products__product-form-variation-table">
+					<thead>
+						<tr>
+							<th></th>
+							<th></th>
+							<th>{ translate( 'Price' ) }</th>
+							<th>{ translate( 'Dimensions & weight' ) }</th>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: Hi! I've found a possible matching string that has already been translated **77** times:
`translate( 'Dimensions' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-variation-rows)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-variation-types-form.js [hunk 7]
```diff
@@ -97,19 +97,19 @@ export default class ProductVariationTypesForm extends Component {
...
+						'The values would be the colors the product is available in.',
 						{ components: { em: <em /> } }
 					) }
 				</p>
 
 				<div className="products__variation-types-form-group">
 					<div className="products__variation-types-form-labels">
 						<FormLabel className="products__variation-types-form-label">{ i18n.translate( 'Variation type' ) }</FormLabel>
-						<FormLabel>{ i18n.translate( 'Variation values' ) }</FormLabel>
+						<FormLabel>{ i18n.translate( 'Values' ) }</FormLabel>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-variation-types-form.js]: Hi! I've found a possible matching string that has already been translated **90** times:
`translate( 'Value' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+							name="width"
+							placeholder={ translate( 'W' ) }
+							type="number"
+							value={ dimensions.width || '' }
+							onChange={ this.setDimension }
+							className="products__product-dimensions-width"
+						/>
+						<FormTextInputWithAffixes
+							name="height"
+							placeholder={ translate( 'H' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **8** times:
`translate( 'h' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+						value={ variation.regular_price || '' }
+						onChange={ this.setPrice }
+						size="4"
+					/>
+				</td>
+				<td className="products__product-dimensions-weight">
+					<div className="products__product-dimensions-input">
+						<FormTextInput
+							name="length"
+							placeholder={ translate( 'L' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'L', { context: 'momentjs format string (sameElse)'} )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+								name="width"
+								placeholder={ translate( 'W' ) }
+								type="number"
+								value={ dimensions.width || '' }
+								onChange={ this.setDimension }
+								className="products__product-dimensions-width"
+							/>
+							<FormTextInputWithAffixes
+								name="height"
+								placeholder={ translate( 'H' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **8** times:
`translate( 'h' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+						onChange={ this.setPrice }
+						size="4"
+					/>
+				</td>
+				<td>
+					<div className="products__product-dimensions-weight">
+						<div className="products__product-dimensions-input">
+							<FormTextInput
+								name="length"
+								placeholder={ translate( 'L' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'L', { context: 'momentjs format string (sameElse)'} )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	toggleStock() {
+		const { variation, product, editProductVariation } = this.props;
+		editProductVariation( product, variation, { manage_stock: ! variation.manage_stock } );
+	}
+
+	// TODO Pull in correct currency, dimension, and weight from from settings.
+	render() {
+		const { variation, translate, manageStock } = this.props;
+		const dimensions = variation.dimensions || { };
+		const allVariationsRow = ! variation.attributes.length;
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: This variable name is confusing.  That being said I do not have a better name.  Reading this PR it took me a minute or two to understand what was going on here.

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+					<FormCurrencyInput
+						currencySymbolPrefix="$"
+						name="price"
+						value={ variation.regular_price || '' }
+						onChange={ this.setPrice }
+						size="4"
+					/>
+				</td>
+				<td>
+					<div className="products__product-dimensions-weight">
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: The indentation gets pretty deep here.  Would it make sense to push some of this into another component?

File: client/extensions/woocommerce/app/products/product-form-variations-card.js [hunk 14]
```diff
@@ -9,9 +9,10 @@ import { localize } from 'i18n-calypso';
  */
 import FoldableCard from 'components/foldable-card';
 import ProductVariationTypesForm from './product-variation-types-form';
+import ProductFormVariationsTable from './product-form-variations-table';
 import FormToggle from 'components/forms/form-toggle';
 
-const ProductFormVariationsCard = ( { product, editProduct, translate, editProductAttribute } ) => {
+const ProductFormVariationsCard = ( { product, variations, editProduct, translate, editProductAttribute, editProductVariation } ) => {
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variations-card.js]: This line is getting really long

File: client/extensions/woocommerce/app/products/product-form-variation-row.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+								value={ variation.weight || '' }
+								onChange={ this.setWeight }
+								size="4"
+							/>
+						</div>
+					</div>
+				</td>
+				<td>
+					<div className="products__product-manage-stock">
+						{ allVariationsRow && (
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Forgive my ignorance.  Is this style of conditional templating a style convention within Calypso.  I've never used this before but it makes sense to me.

File: client/extensions/woocommerce/app/products/product-form-variations-table.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	static propTypes = {
+		variations: PropTypes.array,
+		product: PropTypes.object,
+		editProductVariation: PropTypes.func.isRequired,
+	};
+
+	renderVariationRow( variation ) {
+		const { product, variations, editProductVariation } = this.props;
+		const id = isNumber( variation.id ) && variation.id || 'index_' + variation.id.index;
+		const manageStock = variations && variations[ 0 ].manage_stock || false;
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: why use the first variation and not the variation passed?

Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: if one is manage_stock are they all?

File: client/extensions/woocommerce/style.scss [hunk 17]
```diff
@@ -104,4 +103,110 @@
...
+
+			.products__product-dimensions-input {
+				margin-right: 8px;
+
+				.products__product-dimensions-width,
+				.products__product-dimensions-height {
+					margin-left: -1px;
+				}
+
+				.products__product-dimensions-length,
```
Comment (belcherj) [inline client/extensions/woocommerce/style.scss]: Not sure if it is a calypso rule but it is generally considered a bad idea to nest more than 3 deep in sass.

File: client/extensions/woocommerce/style.scss [hunk 18]
```diff
@@ -104,4 +103,110 @@
...
+				border: 1px solid lighten( $gray, 20% );
+				display: flex;
+
+				input[type=checkbox] {
+					float: none;
+					margin: 0;
+				}
+
+				+ .form-text-input {
+					transform: translateX(-1px);
```
Comment (belcherj) [inline client/extensions/woocommerce/style.scss]: is spacing required inside of parens?

File: client/extensions/woocommerce/style.scss [hunk 19]
```diff
@@ -104,4 +103,110 @@
...
+		overflow-x: auto;
+	}
+
+	/* TODO: remove placeholder styles once implemented */
+	.products__product-form-variation-image {
+		background: $gray-light;
+		height: 40px;
+		width: 40px;
+		margin-bottom: 16px;
+		@include breakpoint( ">960px" ) {
```
Comment (belcherj) [inline client/extensions/woocommerce/style.scss]: single quotes:  https://github.com/Automattic/wp-calypso/blob/master/.stylelintrc#L18

File: client/extensions/woocommerce/app/products/product-form-variations-table.js (hunk 16 above)
Comment (justinshreve) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: The first variation that we generate is one that contains fall back configuration for the other variations. And we decided that you wouldn't manage stock for one variation and not do it for another - so the checkbox for this is supposed to apply to all.

File: client/extensions/woocommerce/app/products/product-form-variation-row.js (hunk 15 above)
Comment (justinshreve) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: I've seen it in other places around the code base and we are using it within the WooCommerce extension dir.

File: client/extensions/woocommerce/app/products/product-form-variations-table.js (hunk 16 above)
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: Works for me! Hopefully I'll be up to speed soon. In the mean time please.excuse all my questions.

Comment (justinshreve) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: No worries! Thanks for providing some feedback already. I'll be working on splitting this branch up into smaller PRs tomorrow and will doing some cleanup using your notes.

File: client/extensions/woocommerce/app/products/product-form-variations-table.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+		const { variations, translate } = this.props;
+		return (
+			<div className="products__product-form-variation-table-wrapper">
+				<table className="products__product-form-variation-table">
+					<thead>
+						<tr>
+							<th></th>
+							<th></th>
+							<th className="products__product-price">{ translate( 'Price' ) }</th>
+							<th>{ translate( 'Dimensions & weight' ) }</th>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-form-variations-table.js]: Hi! I've found a possible matching string that has already been translated **77** times:
`translate( 'Dimensions' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-variation-rows)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-variation-row.js (hunk 13 above)
Comment (justinshreve) [inline client/extensions/woocommerce/app/products/product-form-variation-row.js]: Done. See #13767.

**Comment by kellychoffman** [comment]
Okay, I added horizontal scrolling to the table and am . The dimension and affix components still look unstyled and will be until #13767 is merged.

**Comment by justinshreve** [comment]
> The dimension and affix components still look unstyled and will be until #13767 is merged.

I rebased this PR so it is styled now.

**Comment by justinshreve** [comment]
@allendav @belcherj I think this is ready for a proper review.

Thanks a bunch to @kellychoffman and @jameskoster for making this look so awesome.

**Comment by belcherj** [comment]
@kellychoffman When adding a variation the "Variation values" does not specify how to add multiple values.  Should we change the copy and down the road look at a different UI.

**Comment by justinshreve** [comment]
@belcherj That's actually explained by a missing placeholder. See #13831.

**Comment by belcherj** [comment]
I deleted the comment as I figured out you had to hit enter, thanks @justinshreve

**Pull request merged**